### DISCUSSION TITLE
RAG is not working except actually sometimes it does.

### DISCUSSION THREAD

**Discussion by timnetworks** [open-webui/open-webui#3868] [Q&A] [answered]
Hi all, I'm not sure where to turn with this, as _sometimes it works_. But it really mostly _doesn't_. I got it to read a plaintext file, it repeated about two dozen lines correctly. I checked inside the docker container and files are being uploaded to /backend/app/data/ and processed with a hash value in the beginning of the filename. I also believe I got web retrieval to work yesterday while testing, this may or may not matter.

I had some Tools downloaded from openwebui, removing them did not resolve anything. I had tried different models (gemma2b, mistral7b, phi3-mini, and just now llama3. None of this seems to make difference. I changed some settings in the Documents section of Admin Settings, cleared and created new vectors, didn't seem to make a difference. 

**Ollama**: _(installed via curl sSL I'll be honest)_
● ollama.service - Ollama Service
     Loaded: loaded (/etc/systemd/system/ollama.service; enabled; vendor preset: enabled)
     Active: active (running) since Sun 2024-07-14 15:21:46 UTC; 4h 19min ago
   Main PID: 849 (ollama)
      Tasks: 32 (limit: 38170)
     Memory: 13.5G
        CPU: 13min 21.663s
     CGroup: /system.slice/ollama.service
             └─849 /usr/local/bin/ollama serve
```
Jul 14 19:29:28 tim-srv-infer ollama[849]: llama_new_context_with_model:  CUDA_Host  output buffer size =     2.02 MiB
Jul 14 19:29:28 tim-srv-infer ollama[849]: llama_new_context_with_model:      CUDA0 compute buffer size =   560.00 MiB
Jul 14 19:29:28 tim-srv-infer ollama[849]: llama_new_context_with_model:  CUDA_Host compute buffer size =    24.01 MiB
Jul 14 19:29:28 tim-srv-infer ollama[849]: llama_new_context_with_model: graph nodes  = 1030
Jul 14 19:29:28 tim-srv-infer ollama[849]: llama_new_context_with_model: graph splits = 2
```

**Docker**: _(installed via docker wiki instructions)_
sudo docker run -d --network=host -v open-webui:/app/backend/data -e OLLAMA_BASE_URL=http://172.17.0.1:11434 --name open-webui --restart always ghcr.io/open-webui/open-webui:main

CONTAINER ID   IMAGE                                COMMAND           CREATED        STATUS                 PORTS     NAMES
003dc458e24d   ghcr.io/open-webui/open-webui:main   "bash start.sh"   43 hours ago   Up 2 hours (healthy)             open-webui

[docker_logs_2024_07_14.txt](https://github.com/user-attachments/files/16228617/docker_logs_2024_07_14.txt)

**ComfyUI**: _(installed via venv / git clone)_
```
Total VRAM 8101 MB, total RAM 32005 MB
pytorch version: 2.3.1+cu121
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce GTX 1080 : cudaMallocAsync
Using pytorch cross attention

Import times for custom nodes:
   0.0 seconds: /media/inf/ComfyUI/custom_nodes/websocket_image_save.py
   0.1 seconds: /media/inf/ComfyUI/custom_nodes/ComfyUI-Manager

Starting server
To see the GUI go to: http://0.0.0.0:8188
```

There is a similar ticket [open and closed](https://github.com/open-webui/open-webui/discussions/3843) here, but I'm not understanding what solution was arrived at, just noticing the same symptoms. Attached below is a screenshot of the model getting caught unawares with the same text file it previously read back to me:

![image](https://github.com/user-attachments/assets/8e87668f-01ab-4f8c-af09-e5d0fcdb1ba1)

Thank you in advance,

**Comment by silentoplayz** [comment]
Could you test a few things for me?

1) Determine if RAG works in any chat after the first message that YOU send for a large language model to process.

Example:
Say anything to the LLM, await for a response, THEN use `#` to pick a document and ask a question about it. Record the result and let me know if RAG decides to work on this attempt. If it did, regenerate the LLM's response a few times to ensure this isn't a fluke.

2) Determine if the same behavior of RAG only working *sometimes*, also occurs with external models (Such as GroqCloud models or whatever Open AI API compatible connections you've got for external LLMs) within Open WebUI.

3) Downgrade to v0.2.0 of Ollama (download the OllamaSetup.exe file from their releases page on GitHub) and see if your RAG issues are fixed outright.

**Reply by timnetworks** [reply]
Thanks so much for checking on this! This is for a home lab, I am just experimenting while learning this. I am running this on an 8GB 1080 but also have another 11GB 1080Ti I can use for this purpose if needed. I would rather everything stays contained and can be popped open in a Kiwix server or something.

1. That seems to do the trick! I suppose I'm initializing the model first, then feeding it the data via chat upload, I've attached a log of it working: [rag-chat-log-2024-07-14.txt](https://github.com/user-attachments/files/16228776/rag-chat-log-2024-07-14.txt) I think I read elsewhere this was preferred to adding it to the Knowledge Base. ~~However, uploading the next document into the same chat again reverts it to the original behaviour.~~ I think I'm just going too heavy on the context. With simple files (one or two sentence .txt) it works just fine.

    ![image](https://github.com/user-attachments/assets/1b54f4a6-b5c9-4e5a-b939-97fc8a11189f)


2. I am not using any external calls as far as I understand (and prefer)

3. $ ollama -v
    ollama version is 0.2.2

> Downgrade to v0.2.0

Is this the recommended course of events given the responses above? Thank you,

**Reply by timnetworks** [reply]
@silentoplayz said
> Example:
> Say anything to the LLM, await for a response, THEN use # to pick a document and ask a question about it. 

I think this is the solution for me. Perhaps setting a banner to remind of this. Is this the expected behaviour or is it likely to be patched away?

**Reply by silentoplayz** [reply]
> @silentoplayz said
> 
> > Example:
> > Say anything to the LLM, await for a response, THEN use # to pick a document and ask a question about it.
> 
> I think this is the solution for me. Perhaps setting a banner to remind of this. Is this the expected behaviour or is it likely to be patched away?

This is a bug that has introduced itself into Open WebUI with Ollama v0.2.2 and later versions of Ollama. There should be no need to say something to the LLM first before RAG works in a chat. This bug is causing RAG to not work as intended, and it occurs your bug report might be up the same vine of what I had issues with; hence the 3 questions I provided you with for potential diagnosis of the issue.

Personally, I do recommend attempting to downgrade your version of Ollama. Upon a successful downgrade of Ollama. RAG *should* just **work** again as expected within Open WebUI. There is no ETA on when this issue will be patched out, as there's not enough reports or sufficient information for the Ollama team to go off of right now.

Ollama v0.2.0 can be found here: https://github.com/ollama/ollama/releases/tag/v0.2.0

**Reply by timnetworks** [reply]
Thank you for the detailed rundown on this. I see that feature requests aren't in short supply here so I may just keep it in mind as I move forward on the project. 

I wanted to add lastly that there was a _healthy_ dose of user error here, when the input is structured as if the attachment were just text typed above, the model has no problem picking up context.

![image](https://github.com/user-attachments/assets/7969d069-b162-4214-a59e-c78f13185eb0)

**Reply by qiulang** [reply]
@silentoplayz your solution does not work for me, can you take a look at my question https://github.com/open-webui/open-webui/discussions/5002 ?

Thanks!

**Comment by zouzeTG** [comment]
I have exactly the same problem ! I have download ollama 2.0

**Comment by samyan** [comment]
The same here.

**Comment by nitish-raj** [comment]
I am facing the same issue on `ollama version 0.3.6`

**Comment by gaetanquentin** [comment]
the same with 0.3.9

**Comment by BennisonDevadoss** [comment]
I'm encountering an issue where querying a single document works as expected. However, when selecting multiple documents, the response only considers the most recently added document, ignoring the rest.

What might be causing this behavior? I think this error could be related to RAG.

```
Ollama Version: 0.1.39
```

**Comment by qiulang** [comment]
We probably need to look into the code to see what happen.  Just like @BennisonDevadoss and what I mentioned in my question https://github.com/open-webui/open-webui/discussions/5002  RAG only works when querying a single document.

I check the code and find [queryDoc](https://github.com/open-webui/open-webui/blob/main/src/lib/apis/rag/index.ts#L299C14-L299C22) calls the RAG api in backend. But I don't know Svelte so I can't find out how queryDoc is called. When I search `queryDoc` in src folder, I can't find any reference to it. So can someone knows Svelte help ?

**Reply by thiswillbeyourgithub** [reply]
I'm doing a PR right now please hold

edit: see #5378

**Reply by rcalv002** [reply]
Great work on that, you're a hero to men

**Reply by rcalv002** [reply]
I checked out your PR and it does seem to be pulling appropriate data, and referencing only the document where it actually found relevant content. however right before it starts the embedding, i see this error in the console. sharing in case its relevant..

```
ERROR [open_webui.apps.rag.utils] Error when querying the collection with hybrid_search: 'dict' object has no attribute 'documents'
Traceback (most recent call last):
  File "/Users/rcalv002/AIStuff/open-webui/backend/open_webui/apps/rag/utils.py", line 216, in query_collection_with_hybrid_search
    result = query_doc_with_hybrid_search(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rcalv002/AIStuff/open-webui/backend/open_webui/apps/rag/utils.py", line 135, in query_doc_with_hybrid_search
    raise e
  File "/Users/rcalv002/AIStuff/open-webui/backend/open_webui/apps/rag/utils.py", line 100, in query_doc_with_hybrid_search
    texts=result.documents,
          ^^^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'documents'
```

**Reply by thiswillbeyourgithub** [reply]
Thanks a lot.

I checked an I see nowhere a line `texts=result.documents` anywhere in my local repo nor in the github repo. So I conclude that it might not be relevant to this PR.

Maybe something to do with tools/filters/pipes/pipelines ?

I'd say wait for it to crash and report back :laughing: !

**Reply by rcalv002** [reply]
Very strange. I cloned dev branch from owui and then added your remote and merged it, spun up a fresh install, without any other settings. not sure where the error would be coming from then. I also did more testing and it seems in one convo i can ask something.. it identifies the first document, ask something else pertinent to the second document, it says it cant answer because it only has info about the first document... if i look at the context passed with the second prompt, its including the file from the first question only. Actually even if i start a new chat and ask something relating to the second document, it still only talks about the first.. The only way ive gotten around it is going to the model, removing the "all documents" tag, saving, coming back adding each doc individually

**Reply by thiswillbeyourgithub** [reply]
I was not using the latest dev branch but the latest master now that I think of it. I don't think there's a conflict anyway.

The issue about the number of message is another issue, can't link on mobile but it's not about my PR

**Comment by tjbck** [comment, marked as answer]
In case anyone comes across this in the future, the issue here is most likely caused by Ollama model not having enough context length (default: 2048) for the retrieval context. There isn't a good fix for this other than you modifying the context length from Workspace > Models > Edit Model > Advanced Params > Context length. For troubleshooting, I'd recommend you try out the RAG feature using OpenAI models and see if they work as intended, which, most of the time if not all the time, you'll observe it indeed works as intended. 

In case if it's still not working with external models also, Your Open WebUI most likely have failed to extract content from your document file, which might not have been in the support doc file format list. In either case, RAG pipeline is working as intended here. Hope this helps.

**Comment by qiulang** [comment]
@tjbck I can verify that in my case (Chinese document) modify the context length to 4096 indeed works! Thanks.

But the other factor that affects the result is Top K value, I have tried 3, 5, 6, 8, 10 (I have more than 10 documents) and I find 8+ values work and less than 6 normally does not work well.

**Reply by thiswillbeyourgithub** [reply]
Which embedding model are you using ?

**Reply by qiulang** [reply]
I have tried 8 different embedding models, for Chinese documents, "BAAI/bge-large-zh-v1.5" work best in my case.

1. "all-MiniLM-L12-v2"
2. "BAAI/bge-small-en-v1.5"
3. "distiluse-base-multilingual-cased-v1"
4. "paraphrase-multilingual-MiniLM-L12-v2"
5. "BAAI/bge-small-zh-v1.5"
6. "BAAI/bge-base-zh-v1.5"
7. "BAAI/bge-large-zh-v1.5"
8. "shibing624/text2vec-base-chinese"

I also find that the chunk size and overlap size will affect my result. But I have not found some good combination.

**Comment by qiulang** [comment]
@tjbck I don't understand why top_k value significantly affects the retrieve results in RAG as I think it shouldn't, can you take a look at my question here https://github.com/open-webui/open-webui/discussions/6210.

Besides, I think it makes more sense if we can have one top_k value for retrieval and another top_k value for rerank.

**Comment by danielrosehill** [comment]
I have to say that I've noticed the same thing. Unfortunately, I've played around with all the parameters, experimented between using the local embedding models and OpenAI external ones, and so far haven't cracked the mystery. Not that retrieval is terrible or unusable, but it really does struggle on certain contexts, particularly those that are larger file sizes.

**Reply by Schwenn2002** [reply]
I have the following comments on optimizing the RAG.

Firstly, everyone should **replace** the internal **chromaDB with qdrant**. This already provides a much better document retrieval. There is a simple Docker setup for qdrant.

Secondly, for **larger chunk sizes** (larger than 512 tokens), an embedding model should be selected that also supports a corresponding sequence length (e.g. Snowflake/snowflake-arctic-embed-l-v2.0). 

Thirdly, the above suggestion for two top k should be implemented (I changed it hard-coded) to query documents.:

> The first step is to find top k=30 similar documents in the RAG, these are sorted according to hit probability after reranking. 
> After that, only the best hits top k2=10 are returned from this list.

This can significantly reduce the context window of the LLM and irrelevant chunks are cut off after reranking.

For smaller chunk sizes and less hardware, an overlap of 50% makes sense (256/128 tokens). For larger chunks, I have found that less overlap is sufficient (1024/192 tokens). However, it always depends on the type of document.

**Comment by zakweb3** [comment]
I have monitored the network traffic between OWUI and LLM using Chrome DevTools.

When prompting with an attached document from the knowledge base (KB), OWUI sends the relevant data chunks related to my initial query to the LLM.

However, when I prompt a second query related to another section further down in the same document, the LLM cannot find the answer in the current context and replies with something like, "I do not have the relevant document."

I have tested this with multiple LLMs, including deepseek r1 reasoning, Gemini 2.0 Flash, and gemini-2.0-flash-exp.

Regardless of how I adjust my prompt or RAG settings, OWUI does not send any more data chunks to the LLM after the initial query unless I attach another document in the chat session.

This issue is not related to context length (Gemini 2.0 Flash has a 1 million token context length), chunk size, re-ranking, topK, or any similar parameter.

The desired behavior is that if the LLM cannot find the answer in the current context, OWUI should initiate another RAG retrieval with the second query and send the relevant data chunks to the LLM.

OWUI's RAG appears to be effective at retrieving relevant data for the initial query, which minimize token consumption within the context size. However, it should be able to repeat this process as needed throughout the chat session.

I am unsure whether to open an issue based on these findings.

@tjbck @rcalv002 @silentoplayz

I look forward to your feedback and advice.

Thanks,
Zak.

**Reply by spammenotinoz** [reply]
Try "Full Context Mode", seems much more reliable.
Also if you are using Tika, try this image lexpredict/tika-server has a fix for the space issue that impacts RAG.

**Reply by zakweb3** [reply]
Yes, I know "Full Context Mode" works well for small documents. 
However, consider a 5,000-page document where you need to prompt your LLM about different sections. 
Sending the entire document isn't feasible because it would exceed the context window. 
The core idea of a RAG process is to send only the data chunks relevant to your query to the LLM. 
The problem is that OWUI's RAG implementation seems to do this only for the initial query. 
The RAG process should be re-initiated for subsequent queries focusing on different sections of the document. 
I hope this clarifies the issue.

>Also if you are using Tika, try this image lexpredict/tika-server has a fix for the space issue that impacts RAG.

Noted

**Reply by soq2000** [reply]
Dear All,
I just try to use the RAG feature in OWUI, and I can see that it does not provide a good response as expected. I found some setting tricks from the above discussions. However according to the last one from @zakweb3 , I understand that in any case the OWUI will fail in responding from the second request. If it is true, how should I try to build a RAG chat with OWUI please ?

During some trials with RAG in OWUI, I also faced some concrete issues as the following:
1- Just get lost among a lot of embedding models. From the Perplexity  recommendation , is the E5-large-v2 the best open-source one ?
2- Sometimes while feeding a pdf document, I got the error  400: 'NoneType' object is not iterable (see in the attachment) what can cause such problem?
![image](https://github.com/user-attachments/assets/59c9b63b-7761-46e2-b73d-dd394f9f599c)

Thank you in advance for your help
Best regards

**Reply by zakweb3** [reply]
Q1 using local embedding model can be resource intensive depending on your hardware and the size of your knowledge base. I almost crashed my vps downloading a large embedding model which took up all the disk free space.  
Q2 I got that error sometime due to rate limiting using external embedding model, increasing the embedding batch size fixed the issue for me. I can't tell you exactly whats the issue for you though. Keep in mind you need to reset the vector data base switching embedding model.

**Reply by soq2000** [reply]
Thanks @zakweb3 for your answer. I will try to increase the embedding batch size to see. 
I would like to ask how I can add paid models to the LLM model list so that I can select it as the base model for my RAG chatbot (I do have OpenAI token). Also do you which embeddings model OpenAI uses for theire PlayGround Assistant ? I test the embeddings documents there, it seems that the answers were perfectly matched. I just wonder how they set/adjust the RAG chatbot to give an amazing results. Thanks

**Reply by zakweb3** [reply]
If I understand your question correctly, you just need to add a new OpenAI connection in the admin panel under **Settings > Connections > Manage OpenAI API connection**. Click the '+' sign, then fill in the OpenAI URL, token, and model ID. After completing the details, click the '+' sign again to add the model ID, and then click **Save**. Do not forget to click the **Save** button once more in the bottom right corner to confirm the changes. 

Once this is done, you should find your model listed in the **Models** section of the admin panel.
![image](https://github.com/user-attachments/assets/cf19f418-6079-4318-95c0-c347cc0ffce1)

**Comment by chrisoutwright** [comment]
<img width="1904" height="679" alt="image" src="https://github.com/user-attachments/assets/c5a87fff-ecbe-422e-8a76-0cd1a6aff89b" />
Sometimes I have results with huge files 
https://raw.githubusercontent.com/NixOS/nixpkgs/0c408a087b4751c887e463e3848512c12017be25/pkgs/top-level/all-packages.nix
where suddenly the context window is exceeded and I will get gibberish  .. is it possible to exclude like preventing massive context while parsing the results?

I am not sure what to set exactly:
<img width="1712" height="889" alt="image" src="https://github.com/user-attachments/assets/bbb88103-56a1-4842-90ff-591ca4522356" />


but the k values are not helping in this. Also in search engine (I use searxng and max results to a count of 4  with (5 concurrently)., will not prevent that. I cannot believe there are so many relevant chunk matches there on my query.

What can be done on preventing on some queries the massive information overflow? Is this like a chunking issue or why so much at once?

**Reply by chrisoutwright** [reply]
just to give context :  the 8 website had a context amassed of 600k  it seems for the llm (I had checked in ollama serve .. and the gpu literally kept churning away even though it tried trimming the 600k to my 110k context size  (the trimming seems completely broken in ollama, the gpu will just hang..)

[a comment by an automated account is not shown]

**Comment by stepankotyk** [comment]
it can READ files only if you attach it to FIRST msg. if you send message and then attach file in subsequent messages - it will not see it.  I dont know why it was not fixed for so long, but i spent hours playing with RAG configurations, backends, models only to find out it was  FrontEnd bug :D
